### PR TITLE
URL Clickabillity. OOC Notes/Flavor Text/Species lore

### PR DESCRIPTION
**Pull request by AnywayFarus**
## About The Pull Request

This PR adds the highlight and clickability to URLs in flavor/ooc notes/species lore

## How This Contributes To The Skyrat Roleplay Experience

The random CI inclusions will probably disappear, easier to open links left by players.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>

### Intial Idea.

https://github.com/Skyrat-SS13/Skyrat-tg/assets/78199449/a9786ba6-3d62-4865-9a24-e8807b17fb18

### Changed idea.

NVIDIA_Share_22 10 12 12 номер 400

</details>

## Changelog

:cl: Vishenka0704
qol: In OOC Notes/Flavor Text/Species lore url now are clickable
/:cl:

### CODE REVIEW

**Comment by nikothedude** [comment]
you sure this is a good idea? giving people uncontrolled access to hyperlinks like this is just _begging_ for IP loggers, download links, etc.

**Comment by AnywayFarus** [comment]
> you sure this is a good idea? giving people uncontrolled access to hyperlinks like this is just _begging_ for IP loggers, download links, etc.

Actually, there is a slight concern, but if you think about it, the link with loggers and file downloads can be inserted like that.
And in general, I don't really understand the fear of an IP logger....

**Comment by nikothedude** [comment]
> > you sure this is a good idea? giving people uncontrolled access to hyperlinks like this is just _begging_ for IP loggers, download links, etc.
> 
> Actually, there is a slight concern, but if you think about it, the link with loggers and file downloads can be inserted like that. And in general, I don't really understand the fear of an IP logger....

i dunno what you mean by the first bit but

ip loggers are generally harmless if no identifying data is passed with them, the most you get is "a guy w/ this ip accessed this page"

but lets say you say "hey bro check out my sweet ftext". they click the link, and wham, you have a general idea of where they live, what ISP they use, and, most importantly, you know their IP so you can ddos them
download links are just a god-awful idea in general as well because some people are fucking idiots (or just ulucky and accidentally click things) and install malware onto their PC

**Comment by AnywayFarus** [comment]
> > > you sure this is a good idea? giving people uncontrolled access to hyperlinks like this is just _begging_ for IP loggers, download links, etc.
> > 
> > 
> > Actually, there is a slight concern, but if you think about it, the link with loggers and file downloads can be inserted like that. And in general, I don't really understand the fear of an IP logger....
> 
> i dunno what you mean by the first bit but
> 
> ip loggers are generally harmless if no identifying data is passed with them, the most you get is "a guy w/ this ip accessed this page"
> 
> but lets say you say "hey bro check out my sweet ftext". they click the link, and wham, you have a general idea of where they live, what ISP they use, and, most importantly, you know their IP so you can ddos them download links are just a god-awful idea in general as well because some people are fucking idiots (or just ulucky and accidentally click things) and install malware onto their PC

I'll try to add one thing... hopefully it works, and it should essentially give you a way to keep yourself safe....

**Comment by AnywayFarus** [comment]
image

**Comment by nikothedude** [comment]
nah, people just wont read it, nobody ever does

if you want my opinion, i'd suggest adding a whitelist of domains

things like docs.google.com, imgur, or generally just domains we all know won't be much of a danger

even then though i really doubt a maintainer would accept this PR but, hey, I'm not a maintainer, so feel free to ignore me

**Comment by Jolly-66** [comment]
The idea is good but the execution does open the door to malicious intent.

I'm not someone who can levy weight here but personally I'd be against this from a security point of view.

**Comment by AnywayFarus** [comment]
> nah, people just wont read it, nobody ever does
> 
> if you want my opinion, i'd suggest adding a whitelist of domains
> 
> things like docs.google.com, imgur, or generally just domains we all know won't be much of a danger
> 
> even then though i really doubt a maintainer would accept this PR but, hey, I'm not a maintainer, so feel free to ignore me

I'm going to try to get the maintainers opinion on this version for now, starting from the "that if people don't want to read the link, that way they can just paste their own and also then go if they want to".
If they tell me to redo it, I'll do a check on domains.

**Comment by StrangeWeirdKitten** [comment]
I think if someone wanted to ip log something, they could use a tiny URL service regardless and still do it maliciously. Because people still post links.

**Comment by burgerenergy** [comment]
Just make it the raw text, not something you can alias a link into anything else.

Can people still disguise it with tinyurl etc? Yeah, but then the onus is on the person clicking the mysterious link as opposed to being fooled by a fake out.

**Comment by GoldenAlpharex** [comment]
I think having the ability to add aliases for links, while it's nice, is probably not something to give right away. We can start with allowing hyperlinks in those places, but I think the alias should be left alone for now. I don't trust the players quite yet to let them do that, if they have unrestricted access to hyperlink aliases.

**Comment by StrangeWeirdKitten** [comment]
I think it would be fine if it was restricted to discord image links, f-list, etc.

**Comment by AnywayFarus** [comment]
After talking to @GoldenAlpharex , it was decided to change the concept of the update, and give (hopefully temporarily) clickable links, and if everything is good, maybe after that there will be what I wanted originally.
image

File: tgui/packages/tgui/interfaces/ExaminePanel.js [hunk 1]
```diff
@@ -3,6 +3,32 @@ import { Stack, Section, ByondUi } from '../components';
...
+  const parts = [];
+  let regex = /https?:\/\/[^\s/$.?#].[^\s]*/gi;
+  let lastIndex = 0;
+
+  text.replace(regex, (url, index) => {
+    parts.push(text.substring(lastIndex, index));
+    parts.push(
+      <a
+        style={{
+          'color': '#0591e3',
```
Comment (GoldenAlpharex) [inline tgui/packages/tgui/interfaces/ExaminePanel.js:17]: I'm curious, what's behind the choice of forcing the color to always be the same, for the URLs?

**Review by GoldenAlpharex — COMMENTED** [review]
Outside of that, I'd say it's good for a test-merge.

File: tgui/packages/tgui/interfaces/ExaminePanel.js (hunk 1 above)
Comment (AnywayFarus) [reply tgui/packages/tgui/interfaces/ExaminePanel.js:17]: The easiest way is like the decorations below.
You can make a separate style file, but in my opinion it’s too cool for that.

**Comment by CliffracerX** [comment]
Arbitrary URLs make me very uncomfy, _especially_ given everything BYOND is.  I understand the QOL of being able to link stuff, but- I'm not sure this is a good idea.

**Comment by YakumoChen** [comment]
This seems pretty wildly unsafe
How viable is it to use as a vector for malicious code execution?

**Comment by lessthnthree** [comment]
Please check that nothing was missed in sanitizing tags except links.

https://github.com/Skyrat-SS13/Skyrat-tg/issues/24473

**Comment by AnywayFarus** [comment]
> Please check that nothing was missed in sanitizing tags except links.
> 
> #24473

I don't really see how what you've described is possible.
In theory (and I checked it a couple of times) all <> in those input boxes are eaten without my help.
And the place where they could stick those characters is a link, and if those... did that, I doubt they'd have any effect on the rest of the text outside the link.
Are you sure you didn't observe-- just the characters available for insertion?
I saw one character before my PR, and they were clearly used there, all slanted and barely readable.

**Comment by AnywayFarus** [comment]
𝐿𝑖𝑘𝑒 𝑡ℎ𝑖𝑠?
𝕆𝕣 𝕥𝕙𝕚𝕤❔
Oʀ ᴍᴀʏʙᴇ ᴛʜɪs?
𝓞𝓻 𝓶𝓪𝔂𝓫𝓮 𝓽𝓱𝓲𝓼❓

**Review by GoldenAlpharex — APPROVED** [review]
It's been working out well so far, let's go.

**Pull request merged**